File: bindings/scala/src/main/scala/proven/SafeEmail.scala
// SPDX-License-Identifier: PMPL-1.0

package proven

/**
 * Email validation result with parsed components.
 */
sealed trait EmailResult {
  def isOk: Boolean
  def isError: Boolean = !isOk
  def localPart: Option[String]
  def domain: Option[String]
  def error: Option[String]

  def email: Option[String] = for {
    l <- localPart
    d <- domain
  } yield s"$l@$d"
}

object EmailResult {
  case class Ok(local: String, dom: String) extends EmailResult {
    override def isOk: Boolean = true
    override def localPart: Option[String] = Some(local)
    override def domain: Option[String] = Some(dom)
    override def error: Option[String] = None
  }

  case class Error(msg: String) extends EmailResult {
    override def isOk: Boolean = false
    override def localPart: Option[String] = None
    override def domain: Option[String] = None
    override def error: Option[String] = Some(msg)
  }

  def ok(localPart: String, domain: String): EmailResult = Ok(localPart, domain)
  def error(msg: String): EmailResult = Error(msg)
}

/**
 * Safe email validation and manipulation.
 */
object SafeEmail {

  /** Common disposable email domains */
  val DisposableDomains: Set[String] = Set(
    "tempmail.com",
    "throwaway.email",
    "guerrillamail.com",
    "mailinator.com",
    "10minutemail.com",
    "temp-mail.org",
    "fakeinbox.com",
    "trashmail.com",
    "yopmail.com",
    "sharklasers.com",
    "getairmail.com",
    "tempail.com",
    "discard.email",
    "maildrop.cc"
  )

  private val LocalPartPattern = """^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+$""".r
  private val DomainPattern = """^[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?(\.[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?)*$""".r

  /**
   * Validate an email address (basic check).
   */
  def isValid(email: String): Boolean = parse(email).isOk

  /**
   * Parse and validate an email address.
   */
  def parse(email: String): EmailResult = {
    // Check for empty
    if (email.isEmpty) {
      return EmailResult.error("empty_email")
    }

    // Check length
    if (email.length > 254) {
      return EmailResult.error("email_too_long")
    }

    // Check for @ symbol
    val atIndex = email.indexOf('@')
    if (atIndex == -1) {
      return EmailResult.error("missing_at_symbol")
    }

    // Check for multiple @
    if (email.indexOf('@', atIndex + 1) != -1) {
      return EmailResult.error("multiple_at_symbols")
    }

    // Split at @
    val localPart = email.substring(0, atIndex)
    val domain = email.substring(atIndex + 1)

    // Validate local part
    if (localPart.isEmpty) {
      return EmailResult.error("empty_local_part")
    }
    if (localPart.length > 64) {
      return EmailResult.error("local_part_too_long")
    }
    if (localPart.startsWith(".") || localPart.endsWith(".")) {
      return EmailResult.error("local_part_dot_position")
    }
    if (localPart.contains("..")) {
      return EmailResult.error("local_part_consecutive_dots")
    }
    if (LocalPartPattern.findFirstIn(localPart).isEmpty) {
      return EmailResult.error("invalid_local_part_chars")
    }

    // Validate domain
    if (domain.isEmpty) {
      return EmailResult.error("empty_domain")
    }
    if (domain.length > 253) {
      return EmailResult.error("domain_too_long")
    }

    // Domain must have a dot (unless localhost)
    if (!domain.contains('.') && domain.toLowerCase != "localhost") {
      return EmailResult.error("domain_missing_dot")
    }

    // Domain can't start or end with dot or hyphen
    if (domain.startsWith(".") || domain.endsWith(".") ||
        domain.startsWith("-") || domain.endsWith("-")) {
      return EmailResult.error("invalid_domain_format")
    }

    if (DomainPattern.findFirstIn(domain).isEmpty) {
      return EmailResult.error("invalid_domain_chars")
    }

    EmailResult.ok(localPart, domain)
  }

  /**
   * Extract domain from email.
   */
  def getDomain(email: String): Option[String] = parse(email).domain

  /**
   * Extract local part from email.
   */
  def getLocalPart(email: String): Option[String] = parse(email).localPart

  /**
   * Normalize email (lowercase domain, preserve local part case).
   */
  def normalize(email: String): Option[String] = {
    parse(email) match {
      case EmailResult.Ok(local, dom) => Some(s"$local@${dom.toLowerCase}")
      case _ => None
    }
  }

  /**
   * Fully normalize email (lowercase everything).
   */
  def normalizeFull(email: String): Option[String] = {
    parse(email) match {
      case EmailResult.Ok(local, dom) =>
        Some(s"${local.toLowerCase}@${dom.toLowerCase}")
      case _ => None
    }
  }

  /**
   * Check if email is from a disposable service.
   */
  def isDisposable(email: String): Boolean = {
    getDomain(email).exists(d => DisposableDomains.contains(d.toLowerCase))
  }

  /**
   * Check if email is from a specific domain.
   */
  def isFromDomain(email: String, expectedDomain: String): Boolean = {
    getDomain(email).exists(_.toLowerCase == expectedDomain.toLowerCase)
  }

  /**
   * Check if email is from one of a list of allowed domains.
   */
  def isFromAllowedDomain(email: String, allowedDomains: Seq[String]): Boolean = {
    getDomain(email).exists { domain =>
      allowedDomains.exists(_.toLowerCase == domain.toLowerCase)
    }
  }

  /**
   * Get the TLD (top-level domain) of an email.
   */
  def getTld(email: String): Option[String] = {
    getDomain(email).flatMap { domain =>
      val lastDot = domain.lastIndexOf('.')
      if (lastDot == -1) None
      else Some(domain.substring(lastDot + 1).toLowerCase)
    }
  }

  /**
   * Obfuscate email for display (e.g., "u***r@example.com").
   */
  def obfuscate(email: String): Option[String] = {
    parse(email) match {
      case EmailResult.Ok(local, dom) =>
        val obfuscatedLocal = if (local.length <= 2) {
          "***"
        } else {
          s"${local.head}***${local.last}"
        }
        Some(s"$obfuscatedLocal@$dom")
      case _ => None
    }
  }

  /**
   * Check if two emails are equivalent (after normalization).
   */
  def areEquivalent(email1: String, email2: String): Boolean = {
    (normalizeFull(email1), normalizeFull(email2)) match {
      case (Some(n1), Some(n2)) => n1 == n2
      case _ => false
    }
  }
}

File: .UNSAFE-ALL-BINDINGS-DELETED-FINAL-20260125/bindings/clojure/src/proven/safe_email.clj
;; SPDX-License-Identifier: PMPL-1.0
;; SPDX-FileCopyrightText: 2025 Hyperpolymath

(ns proven.safe-email
  "Safe email validation and manipulation."
  (:require [clojure.string :as str]))

(def disposable-domains
  #{"tempmail.com" "throwaway.email" "guerrillamail.com" "mailinator.com"
    "10minutemail.com" "temp-mail.org" "fakeinbox.com" "trashmail.com"
    "yopmail.com" "sharklasers.com" "getairmail.com" "tempail.com"
    "discard.email" "maildrop.cc"})

(def ^:private local-part-pattern #"^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+$")
(def ^:private domain-pattern #"^[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?(\.[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?)*$")

(defn parse
  "Parse and validate an email address.
   Returns {:ok {:local-part x :domain y}} or {:error msg}."
  [email]
  (cond
    (str/blank? email)
    {:error "empty_email"}

    (> (count email) 254)
    {:error "email_too_long"}

    (not (str/includes? email "@"))
    {:error "missing_at_symbol"}

    (> (count (filter #{\@} email)) 1)
    {:error "multiple_at_symbols"}

    :else
    (let [at-index (str/index-of email "@")
          local-part (subs email 0 at-index)
          domain (subs email (inc at-index))]
      (cond
        (str/blank? local-part)
        {:error "empty_local_part"}

        (> (count local-part) 64)
        {:error "local_part_too_long"}

        (or (str/starts-with? local-part ".")
            (str/ends-with? local-part "."))
        {:error "local_part_dot_position"}

        (str/includes? local-part "..")
        {:error "local_part_consecutive_dots"}

        (not (re-matches local-part-pattern local-part))
        {:error "invalid_local_part_chars"}

        (str/blank? domain)
        {:error "empty_domain"}

        (> (count domain) 253)
        {:error "domain_too_long"}

        (and (not (str/includes? domain "."))
             (not= (str/lower-case domain) "localhost"))
        {:error "domain_missing_dot"}

        (or (str/starts-with? domain ".")
            (str/ends-with? domain ".")
            (str/starts-with? domain "-")
            (str/ends-with? domain "-"))
        {:error "invalid_domain_format"}

        (not (re-matches domain-pattern domain))
        {:error "invalid_domain_chars"}

        :else
        {:ok {:local-part local-part :domain domain}}))))

(defn valid?
  "Check if email is valid."
  [email]
  (contains? (parse email) :ok))

(defn get-domain
  "Extract domain from email. Returns nil if invalid."
  [email]
  (when-let [result (:ok (parse email))]
    (:domain result)))

(defn get-local-part
  "Extract local part from email. Returns nil if invalid."
  [email]
  (when-let [result (:ok (parse email))]
    (:local-part result)))

(defn normalize
  "Normalize email (lowercase domain, preserve local part case).
   Returns nil if invalid."
  [email]
  (when-let [result (:ok (parse email))]
    (str (:local-part result) "@" (str/lower-case (:domain result)))))

(defn normalize-full
  "Fully normalize email (lowercase everything).
   Returns nil if invalid."
  [email]
  (when-let [result (:ok (parse email))]
    (str (str/lower-case (:local-part result)) "@" (str/lower-case (:domain result)))))

(defn disposable?
  "Check if email is from a disposable service."
  [email]
  (when-let [domain (get-domain email)]
    (contains? disposable-domains (str/lower-case domain))))

(defn from-domain?
  "Check if email is from a specific domain."
  [email expected-domain]
  (when-let [domain (get-domain email)]
    (= (str/lower-case domain) (str/lower-case expected-domain))))

(defn from-allowed-domain?
  "Check if email is from one of a list of allowed domains."
  [email allowed-domains]
  (when-let [domain (get-domain email)]
    (let [lower-domain (str/lower-case domain)]
      (some #(= (str/lower-case %) lower-domain) allowed-domains))))

(defn get-tld
  "Get the TLD (top-level domain) of an email."
  [email]
  (when-let [domain (get-domain email)]
    (let [last-dot (str/last-index-of domain ".")]
      (when last-dot
        (str/lower-case (subs domain (inc last-dot)))))))

(defn obfuscate
  "Obfuscate email for display (e.g., 'u***r@example.com')."
  [email]
  (when-let [result (:ok (parse email))]
    (let [local (:local-part result)
          domain (:domain result)
          obfuscated-local (if (<= (count local) 2)
                            "***"
                            (str (first local) "***" (last local)))]
      (str obfuscated-local "@" domain))))

(defn equivalent?
  "Check if two emails are equivalent (after normalization)."
  [email1 email2]
  (let [n1 (normalize-full email1)
        n2 (normalize-full email2)]
    (and n1 n2 (= n1 n2))))
